File: content/snippets/go/design_patterns_builder.go
package main

import "fmt"

type Report struct {
	Title    string
	Sections []string
}

type ReportBuilder struct {
	report Report
}

func (b *ReportBuilder) Title(value string) *ReportBuilder {
	b.report.Title = value
	return b
}

func (b *ReportBuilder) Section(value string) *ReportBuilder {
	b.report.Sections = append(b.report.Sections, value)
	return b
}

func (b *ReportBuilder) Build() Report {
	return b.report
}

func main() {
	report := (&ReportBuilder{}).
		Title("Daily Build").
		Section("tests: green").
		Section("deploy: staged").
		Build()

	fmt.Println(report.Title)
	for _, section := range report.Sections {
		fmt.Println("-", section)
	}
}

File: content/snippets/typescript/design_patterns_builder.ts
type Report = {
  title: string;
  sections: string[];
};

class ReportBuilder {
  private report: Report = { title: '', sections: [] };

  title(value: string): this {
    this.report.title = value;
    return this;
  }

  section(value: string): this {
    this.report.sections.push(value);
    return this;
  }

  build(): Report {
    return this.report;
  }
}

const report = new ReportBuilder()
  .title('Daily Build')
  .section('tests: green')
  .section('deploy: staged')
  .build();

console.log(report.title);
for (const section of report.sections) {
  console.log(`- ${section}`);
}